### DISCUSSION TITLE
Collection got reset

### DISCUSSION THREAD

**Discussion by LOKSTED** [ppy/osu#13350] [Q&A] [answered]
[runtime.log](https://github.com/ppy/osu/files/6603857/runtime.log)
[updater.log](https://github.com/ppy/osu/files/6603858/updater.log)
[database.log](https://github.com/ppy/osu/files/6603859/database.log)
[performance.log](https://github.com/ppy/osu/files/6603862/performance.log)

this may be a rare case
I stopped playing lazer last week to work on a beatmap on stable. I came back to lazer and there was an update notification and I was logged out and my collection was reset. I thought the collection might be tied to my account so I update it and login but no luck.
Beatmaps, settings and scores are all fine. I only manually edit in the AppData\Local\osu!\Songs but only with the song folder I was working on on stable.

Please point me to the file/folder where Lazer keeps track of collection. Maybe I can recover something... thanks

**Comment by frenzibyte** [comment]
That's weird. I'm not sure how an update could result in deleting all your collections, the only way I can see that possible is by you misclicking the `Delete all collections` button in the bottom of the settings? which I don't believe to be the case.

osu!lazer keeps track of everything in database files, can you attach your `collection.db` file? (resides in the osu! folder which can be navigated to by the `Open osu! folder` button in the settings)

**Reply by LOKSTED** [reply]
https://drive.google.com/file/d/1-VYh9jHT5WvCLxYqwky_x1kiXX_Rt2nF/view?usp=sharing

**Reply by smoogipoo** [reply]
Here's your fixed database: [collection.db.zip](https://github.com/ppy/osu/files/6608432/collection.db.zip) Just unzip this over your existing lazer collections.db.

**Reply by LOKSTED** [reply]
> Here's your fixed database: [collection.db.zip](https://github.com/ppy/osu/files/6608432/collection.db.zip) Just unzip this over your existing lazer collections.db.

Thank you very much! Just out of curiosity, how did you fix it?

**Reply by smoogipoo** [reply]
By forcing osu! to read the entire file.

**Comment by smoogipoo** [comment]
Not sure how this is possible, possibly a random cosmic event or something. The collections are present but the count written is 0:

![image](https://user-images.githubusercontent.com/1329837/121009410-19351900-c7cf-11eb-948d-a202a75ab04c.png)

**Reply by peppy** [reply]
It looks like this can very feasibly happen if the save operation occurred during an import or delete:

![20210608 122834 (JetBrains Rider-EAP)](https://user-images.githubusercontent.com/191335/121118626-2563b980-c855-11eb-9120-9b130bcd1cac.png)

Shall we convert this to a tracking issue? I think a `ToArray()` call would be enough to guard against this particular fail case, where an exception isn't thrown but numbers serialised may not match.

**Reply by smoogipoo** [reply]
But if Collections.Clear() happened, wouldn't you also expect there to be no collections?

Put another way, the only way I can see what you're saying to be feasible is if the collections were cleared before the `.Count` call, and then populated after the call, which is the complete opposite interaction of your screenshot, and unlikely based on the user's experience - "updated, then collections reset".

**Reply by smoogipoo** [reply]
Actually, I can see one way my process can happen - if the collection was saved during the BDL load. Will test more...

**Comment by peppy** [comment, marked as answer]
This should have been resolved via https://github.com/ppy/osu/pull/13370.

**Reply by LOKSTED** [reply]
I have a backup now but my collection got reset again. Is there anything I should avoid doing to prevent this?

**Reply by peppy** [reply]
Just to confirm, the game made a backup for you this time?

Do you have logs from when this occurred?

**Reply by LOKSTED** [reply]
The game didn't make a backup. The backup is the zip file I got from @smoogipoo. I noticed that it was gone about 3 hours ago.
[database.log](https://github.com/ppy/osu/files/6738653/database.log)
[network.log](https://github.com/ppy/osu/files/6738649/network.log)
[performance.log](https://github.com/ppy/osu/files/6738650/performance.log)
[runtime.log](https://github.com/ppy/osu/files/6738651/runtime.log)
[updater.log](https://github.com/ppy/osu/files/6738652/updater.log)

Collection.db https://drive.google.com/file/d/1YuU4Euk97JQygji3oH1M_IWHezxpNC0d/view?usp=sharing

Maybe something's wrong with my pc since no one's having the same problem?

**Reply by smoogipoo** [reply]
> I noticed that it was gone

When you say this, what do you mean? Did you notice that the _file_ was gone, or did you open osu! and notice that all of your collections were gone? There's a difference between the two - the file being gone isn't necessarily bad if there's a `collections.db.bak` file sitting in there - that .bak file will be restored on the next run of the game. And if the restoration fails, I believe the game will crash.

Unfortunately the logs don't reveal anything.

Where did you get that `collection.db` from? Is it the same one you posted in the OP? It has the same issue as last time, which is suspicious.

**Reply by LOKSTED** [reply]
It was the file fixed by @smoogipoo. He posted the fixed file 22 days ago but I just got around to using it. I put it in AppData\Roaming\osu then play and added some songs to it. There is no collections.db.bak in AppData\Roaming\osu.

The fixed file was good. It only got corrupted after I play lazer for a few days

**Reply by peppy** [reply]
Next time this happens, we need a copy of the "corrupt" file please.

**Reply by LOKSTED** [reply]
Maybe corrupt is not the right word I just don't know the word to describe it but I have it here
Collection.db https://drive.google.com/file/d/1YuU4Euk97JQygji3oH1M_IWHezxpNC0d/view?usp=sharing

**Reply by peppy** [reply]
Yeah not sure. The file you're providing there should be impossible behaviour on the release of osu! you reported it occurring (as in we made a fix which guarantees the database can never be in the state you provided).

I'm guessing the files got mixed up at some point. Please provide a newer file should it happen to you again.
